pkg/controller/secret/secret_controller.go

```diff
@@ -0,0 +1,170 @@
+	if err != nil {
+		return errors.Wrapf(err, "could not create the controller: %v", ControllerName)
+	}
+	predicateFilter := predicate.Funcs{
+		// ignore create event for userData secret creation
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		// get update event only when secret data is changed
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if string(e.ObjectOld.(*core.Secret).Data["userData"][:]) != string(e.ObjectNew.(*core.Secret).Data["userData"][:]) {
+				return true
+			}
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource UserDataSecret
+	err = c.Watch(&source.Kind{Type: &core.Secret{ObjectMeta: meta.ObjectMeta{Namespace: userDataNamespace,
+		Name: userDataSecret}}}, &handler.EnqueueRequestForObject{}, predicateFilter)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// blank assignment to verify that ReconcileSecret implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileSecret{}
+
+// ReconcileSecret reconciles a Secret object
+type ReconcileSecret struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Secret object and makes changes based on the state read
+// and what is in the Secret.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + userDataSecret)
+
+	userData, err := secrets.GenerateUserData()
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "error generating userData secret")
+
+	}
+	// Fetch UserData instance
+	err = r.client.Get(context.TODO(), kubeTypes.NamespacedName{Name: userData.Name, Namespace: userData.Namespace}, &core.Secret{})
+	if err != nil && k8sapierrors.IsNotFound(err) {
+		// Secret is deleted
+		reqLogger.Info("UserData secret not found, recreating the secret.")
+		err = r.client.Create(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		reqLogger.Error(err, "error retrieving the userData secret")
+		return reconcile.Result{}, err
+	} else {
+		// secret is updated
+		reqLogger.Info("Restoring the userData secret")
+		err = r.client.Update(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret updated successfully
+		return reconcile.Result{}, nil
+	}
+}
+
+// createValidUserData creates userData secret if it is not present in the required namespace.
```

Comment:
I would rename this to `createUserData()`. It being valid is implicit as there is no method to create an invalid one.